File: News_Service_JDBC/src/main/java/com/dunice/projectNews/services/impl/UserServiceImpl.java
package com.dunice.projectNews.services.impl;

import com.dunice.projectNews.dto.request.PutUserRequest;
import com.dunice.projectNews.dto.response.PublicUserResponse;
import com.dunice.projectNews.dto.common.BaseSuccessResponse;
import com.dunice.projectNews.exception.CustomException;
import com.dunice.projectNews.mapper.UserMapper;
import com.dunice.projectNews.models.News;
import com.dunice.projectNews.models.User;
import com.dunice.projectNews.repository.UserRepositoryJDBC;
import com.dunice.projectNews.services.UserService;
import com.dunice.projectNews.util.ServerErrorCodes;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UserServiceImpl implements UserService {

    public final UserRepositoryJDBC userRepository;

    private final UserMapper userMapper;

    @Override
    public List<PublicUserResponse> getAllUser() {
        return userMapper.listUserToListPublicUserResponse(userRepository.getAllUser());
    }

    @Override
    public PublicUserResponse replaceUser(PutUserRequest publicUserView) {
        User user = userRepository
                .findByEmail(SecurityContextHolder.getContext().getAuthentication().getName())
                .orElseThrow(() -> new CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT));
        User userUpdate = userMapper.putUserRequestToUser(publicUserView);
        userUpdate.setUserId(user.getUserId());
        userUpdate.setPassword(user.getPassword());
        userUpdate.setNews(user.getNews());
        userRepository.saveUser(userUpdate);
        return userMapper.userToPublicUserResponse(userUpdate);
    }

    @Override
    @Transactional
    public void deleteUser() {
        userRepository.deleteUserByEmail(SecurityContextHolder.getContext().getAuthentication().getName());
    }

    @Override
    public PublicUserResponse getUserInfoById(UUID id) {
        return userMapper
                .userToPublicUserResponse(userRepository
                        .findById(id)
                        .orElseThrow(() -> new CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT)));
    }

    @Override
    public PublicUserResponse getUserInfo() {
        return userMapper
                .userToPublicUserResponse(userRepository
                        .findByEmail(SecurityContextHolder.getContext().getAuthentication().getName())
                        .orElseThrow(() -> new CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT)));
    }

    public void saveUserForNews(News news) {
        User user =
                userRepository
                        .findByEmail(SecurityContextHolder.getContext().getAuthentication().getName())
                        .orElseThrow(() -> new CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT));
        news.setAuthor(user);
    }

    public Boolean checkingUserCurrentAuthenticationUser(User user) {
        User userAuthentication =
                userRepository
                        .findByEmail(SecurityContextHolder.getContext().getAuthentication().getName())
                        .orElseThrow(() -> new CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT));
        return user.getEmail().equals(userAuthentication.getEmail());
    }
}

File: News_Service_Kotlin/src/main/kotlin/com/dunice/advancedServiceKotlin/services/impl/UserServiceImpl.kt
package com.dunice.advancedServiceKotlin.services.impl

import News
import com.dunice.advancedServiceKotlin.dto.request.PutUserRequest
import com.dunice.advancedServiceKotlin.dto.response.PublicUserResponse
import com.dunice.advancedServiceKotlin.exception.CustomException
import com.dunice.advancedServiceKotlin.mapper.UserMapper
import com.dunice.advancedServiceKotlin.models.User
import com.dunice.advancedServiceKotlin.repository.UserRepository
import com.dunice.advancedServiceKotlin.services.UserService
import com.dunice.advancedServiceKotlin.util.ServerErrorCodes
import jakarta.transaction.Transactional
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserServiceImpl @Autowired constructor(val userRepository: UserRepository, private val userMapper: UserMapper) :
    UserService {
    override val allUser: List<PublicUserResponse?>?
        get() = userMapper.listUserToListPublicUserResponse(userRepository.findAll().filterNotNull())

    override fun replaceUser(publicUserView: PutUserRequest?): PublicUserResponse? {
        val user = userRepository
            .findByEmail(SecurityContextHolder.getContext().authentication.name)
            ?: throw CustomException(ServerErrorCodes.USER_NOT_FOUND)
        val userUpdate = userMapper.putUserRequestToUser(publicUserView  ?: throw CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT))
        userUpdate.userId = user.userId
        userUpdate.password = user.password
        userUpdate.news = user.news
        userRepository.save(userUpdate)
        return userMapper.userToPublicUserResponse(userUpdate)
    }

    @Transactional
    override fun deleteUser() {
        userRepository
            .deleteByEmail(SecurityContextHolder.getContext().authentication.name)
    }

    override fun getUserInfoById(id: UUID): PublicUserResponse? {
        return userMapper
            .userToPublicUserResponse(
                userRepository
                    .findById(id)
                    .orElseThrow { CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT) }
                    ?: throw CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT)
            )
    }

    override val userInfo: PublicUserResponse?
        get() = userMapper
            .userToPublicUserResponse(
                userRepository
                    .findByEmail(SecurityContextHolder.getContext().authentication.name)
                    ?: throw CustomException(ServerErrorCodes.USER_NAME_HAS_TO_BE_PRESENT)
            )

    override fun saveUserForNews(news: News) {
        val user =
            userRepository
                .findByEmail(SecurityContextHolder.getContext().authentication.name)
                ?: throw CustomException(ServerErrorCodes.USER_NOT_FOUND)
        news.author = user
    }

    override fun checkingUserCurrentAuthenticationUser(user: User): Boolean {
        val userAuthentication =
            userRepository
                .findByEmail(SecurityContextHolder.getContext().authentication.name)
                ?: throw CustomException(ServerErrorCodes.USER_NOT_FOUND)
        return user.email == userAuthentication.email
    }
}
